### PR TITLE
feat(eth): verify pre-built eth txns

### PR DESCRIPTION
**Pull request by benjamin-wu-bitgo**
### Feature Description 

Client-side verification of transaction pre-builds received from BitGo servers should be verified to match transaction parameters specified by the client before the client signs the transaction.

This is an extra safeguard to protect the client from blindly signing malicious transaction pre-builds from the BitGo servers that do not match the transaction parameters specified by the client. An example of when this can be an issue is if BItGo servers were to be compromised and pre-builds are getting sent to the client that try to send funds to an attacker instead of the recipient address provided by the client.

### Implementation

This PR implements the `verifyTransaction` function for Eth which accepts `txParams` supplied by the client, `txPrebuild` supplied by the BitGo server, and the `wallet` for the transaction as parameters. The function returns true if the transaction is verified, and fails to verify throwing an error with the reason for failing otherwise.

It first checks that the parameters required exist, that the `txPrebuild` only has one recipient (note even multi-recipient transactions specified by the client `txParams` get squashed into a single recipient send to a batcher contract address with a data payload), and that the `txPrebuild` is not both a hop transaction and a batcher transaction.

Then, depending on the type of `txPrebuild` (hop, batcher, or normal), we perform different verification logic.

For a hop transaction, we check that there is only one recipient in the `txParams` and that the recipient amounts match those in `txParams`.  Then, we verify that the `txPrebuild` actually sends to the hop contract address corresponding to the `hopTransaction` in the `txPrebuild`. Finally, we convert the `txParams` `TransactionRecipient` array into a `Recipient` array by retyping the `amount` field from `string | number` to `string` to leverage the existing `validateHopPrebuild` function which checks the `hopTransaction` sends the correct amount to the recipient specified in `txParams`.

For a batch transaction, we verify that the `txPrebuild` is sending the correct total amount to a batcher contract address.

For a normal transaction (that is neither of the above), we check that there is only one recipient in the `txParams` and that the recipient amounts and address match those in `txParams`.

Lastly, we always check the coin of the `txPrebuild` matches the coin we are sending from the wallet.

### Testing

This PR includes unit tests covering all lines in the `verifyTransaction` function.

### Future Considerations

We do not yet fully verify batch transactions as we do not verify the data payload sent to the batcher contract address dictating how the total funds are re-distributed to the multiple recipients. Batch contracts are not yet in prod and parsing the data payload to verify this would be challenging and not worth doing at the moment. We do verify we are sending to a batcher contract address with the correct total amount though.

### CODE REVIEW

File: modules/core/src/v2/coins/eth.ts
Comment (tylerlevine) [inline modules/core/src/v2/coins/eth.ts:1543]: Could these individual validation cases be broken out into separate `protected` functions (eg, `verifyHopTransaction()`, `verifyBatchTransaction()`, etc)? That should help improve readability.

File: modules/statics/src/coins.ts
Comment (tylerlevine) [inline modules/statics/src/coins.ts:30]: I'm not sure this is a great place for this, but putting it on the coin might be more trouble than it's worth. If you do want to do this though, here's what I'd do:

* Add a `batcherContractAddress?: string` property to the [`EthereumNetwork` interface](https://github.com/BitGo/BitGoJS/blob/db7d1c1819d31632c8d2a89387003f944443362a/modules/statics/src/networks.ts#L47)
* Add the `batcherContractAddress: '0x_the_address'` property to the implementation for Ethereum, Kovan, and Goerli ([here](https://github.com/BitGo/BitGoJS/blob/db7d1c1819d31632c8d2a89387003f944443362a/modules/statics/src/networks.ts#L218), [here](https://github.com/BitGo/BitGoJS/blob/db7d1c1819d31632c8d2a89387003f944443362a/modules/statics/src/networks.ts#L238), and [here](https://github.com/BitGo/BitGoJS/blob/db7d1c1819d31632c8d2a89387003f944443362a/modules/statics/src/networks.ts#L246))

Then you can get the correct batcher contract address for the current coin by capturing the 2nd param to `createInstance` (the statics coin, see [here](https://github.com/BitGo/BitGoJS/blob/5ad86846805b94eec3f125f33a8579286c3fc7d8/modules/core/src/v2/coinFactory.ts#L90). It's currently not being captured in the `createInstance` factory method). If you capture it as `this.staticsCoin`, then you can get the actual batcher address using `this.staticsCoin.network.batcherAddress`.

**Review by tylerlevine — APPROVED** [review]
Nice work, no major problems for me but I have a couple ideas on how to clean things up a bit.

Please wait for a review from someone on the eth team however so they can check that the verification logic is correct and comprehensive.

File: modules/core/src/v2/coins/eth.ts
Comment (wesgraham) [inline modules/core/src/v2/coins/eth.ts:245]: will this field ever be empty? Unsure, if it can be ignore

File: modules/core/src/v2/coins/eth.ts
Comment (wesgraham) [inline modules/core/src/v2/coins/eth.ts:253]: Also might be worth leaving a TODO for the london hard fork task that will likely need to update this

File: modules/core/src/v2/coins/eth.ts
Comment (benjamin-wu-bitgo) [reply modules/core/src/v2/coins/eth.ts:1542]: I considered splitting out the validation cases into separate functions but it seems that doing so would require me to add a bunch of extra null/undefined checks (by pushing the check that the required parameters are provided down into the validation function for each case) or else the linter would complain.

Alternatively, I would need to create new interfaces that mark parameters as required rather than optional (e.g. create a new txPrebuild interface for hop transaction validation that has the hopTransaction parameter as required instead of optional in the txPrebuild).

Either way, it seems like it trying to split out functions for each case will create a lot of extra lines of code.

Unless you have an idea for splitting out functions without having to add extra interfaces or null/undefined checks, this doesn't seem worth it.

File: modules/core/src/v2/coins/eth.ts
Comment (benjamin-wu-bitgo) [reply modules/core/src/v2/coins/eth.ts]: Some of these fields won't ever be empty so I updated the interface to no longer mark them as optional.

File: modules/core/src/v2/coins/eth.ts
Comment (benjamin-wu-bitgo) [reply modules/core/src/v2/coins/eth.ts:253]: Added TODO comment

File: modules/statics/src/coins.ts
Comment (benjamin-wu-bitgo) [reply modules/statics/src/coins.ts]: Moved `batcherContractAddress` to the `EthereumNetwork` interface

**Review by tylerlevine — APPROVED** [review]
lgtm

:1st_place_medal:

**Review by wesgraham — APPROVED** [review]

**Pull request merged**
